### ISSUE TITLE
Issue with AutoMapper in AWS Lambda

### ISSUE THREAD

**Issue by Kralizek** [aws/aws-lambda-dotnet#53]
Hi,

I'm trying to use AutoMapper in a Lambda function to convert DTOs to/from domain entities.

The error I get is the following

```
{
  "errorType": "LambdaException",
  "errorMessage": "An exception was thrown when the constructor for type 'EMG.Transactions.LambdaFunction' was invoked. Check inner exception for more details.",
  "cause":   {
    "errorType": "TargetInvocationException",
    "errorMessage": "Exception has been thrown by the target of an invocation.",
    "stackTrace": [
      "at System.RuntimeTypeHandle.CreateInstance(RuntimeType type, Boolean publicOnly, Boolean noCheck, Boolean& canBeCached, RuntimeMethodHandleInternal& ctor, Boolean& bNeedSecurityCheck)",
      "at System.RuntimeType.CreateInstanceSlow(Boolean publicOnly, Boolean skipCheckThis, Boolean fillCache, StackCrawlMark& stackMark)",
      "at System.Activator.CreateInstance(Type type, Boolean nonPublic)",
      "at System.Activator.CreateInstance(Type type)"
    ],
    "cause":     {
      "errorType": "TargetInvocationException",
      "errorMessage": "Exception has been thrown by the target of an invocation.",
      "stackTrace": [
        "at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)",
        "at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)",
        "at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)",
        "at Microsoft.AspNetCore.Hosting.Internal.ConfigureServicesBuilder.Invoke(Object instance, IServiceCollection exportServices)",
        "at Microsoft.AspNetCore.Hosting.Internal.WebHost.EnsureApplicationServices()",
        "at Microsoft.AspNetCore.Hosting.Internal.WebHost.BuildApplication()",
        "at Microsoft.AspNetCore.Hosting.WebHostBuilder.Build()",
        "at Amazon.Lambda.AspNetCoreServer.APIGatewayProxyFunction..ctor()"
      ],
      "cause":       {
        "errorType": "ArgumentNullException",
        "errorMessage": "Value cannot be null.
Parameter name: assembly",
        "stackTrace": [
          "at Microsoft.Extensions.DependencyModel.DependencyContextLoader.Load(Assembly assembly)",
          "at System.Lazy`1.CreateValue()",
          "at System.Lazy`1.LazyInitValue()",
          "at EMG.Transactions.Startup.ConfigureServices(IServiceCollection services)"
        ]
      }
    }
  }
} 
```

And I was able to detect that this is the offending block

```
            services.AddAutoMapper(mce =>
            {
                mce.AddProfile<Mappings.TransactionProfile>();
                mce.AddProfile<Mappings.BookingProfile>();
            });
```

I got a feeling the issue is somehow related to this one: NancyFx/Nancy#2692

Is there any workaround/fix?

**Comment by bjorg** [comment]
@Kralizek are you able to run it in .NET Core 1.0 on the command line?

**Comment by travisgosselin** [comment]
@Kralizek I also run into the same issue in using the "AddAutoMapper" DI. I was not able to resolve, but I used the following to manually initialize as a workaround - but bypasses the use of profiles:
 Mapper.Initialize(config =>
            {
                config.CreateMap<IService, Service>();

            });

I also added a DI reference for IMapper manually:  
services.AddTransient<IMapper>(provider => Mapper.Instance);

**Comment by Kralizek** [comment]
@bjorg the code runs smoothly on my developer machine. The issue happens only once the function is deployed on Lambda.

@travisgosselin I will try this. Thank you.

**Comment by Kralizek** [comment]
@travisgosselin any specific reason for using Transient instead of Singleton?

**Comment by travisgosselin** [comment]
@Kralizek not at all... just typing to quickly. Definitely more appropriate as a Singleton.

**Comment by normj** [comment]
Looks like AutoMapper is having trouble finding assemblies in Lambda. We'll take a look into this but I was able to work around this passing in the assemblies in the `AddAutoMapper` call.

```
using System.Reflection;
...


services.AddAutoMapper(mce =>
{
    mce.AddProfile<Mappings.TransactionProfile>();
    mce.AddProfile<Mappings.BookingProfile>();
}, typeof(Startup).GetTypeInfo().Assembly);
```

**Comment by normj** [comment]
Closing this in favor issue https://github.com/aws/aws-lambda-dotnet/issues/149 which is the .NET Core 2.0 support where this issue will be fixed for finding assemblies.
